actors/builtin/market/market_test.go

```diff
@@ -693,6 +695,363 @@ func TestPublishStorageDealsFailures(t *testing.T) {
+
+		// provider2 publishes deal4 and deal5
+		provider2 := tutil.NewIDAddr(t, 401)
+		mAddrs.provider = provider2
+		dealId4 := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch, endEpoch)
+		dealId5 := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch, endEpoch+1)
+
+		// provider1 activates deal 1 and deal2 but that does not activate deal3 to deal5
+		actor.activateDeals(rt, sectorExpiry, provider, currentEpoch, dealId1, dealId2)
+		actor.assertDealsNotActivated(rt, currentEpoch, dealId3, dealId4, dealId5)
+
+		// provider3 activates deal5 but that does not activate deal3 or deal4
+		actor.activateDeals(rt, sectorExpiry, provider2, currentEpoch, dealId5)
+		actor.assertDealsNotActivated(rt, currentEpoch, dealId3, dealId4)
+
+		// provider1 activates deal3
+		actor.activateDeals(rt, sectorExpiry, provider, currentEpoch, dealId3)
+		actor.assertDealsNotActivated(rt, currentEpoch, dealId4)
+	})
+}
+
+func TestActivateDealFailures(t *testing.T) {
+	marketActor := tutil.NewIDAddr(t, 100)
+	owner := tutil.NewIDAddr(t, 101)
+	provider := tutil.NewIDAddr(t, 102)
+	worker := tutil.NewIDAddr(t, 103)
+	client := tutil.NewIDAddr(t, 104)
+	mAddrs := &minerAddrs{owner, worker, provider}
+
+	startEpoch := abi.ChainEpoch(10)
+	endEpoch := abi.ChainEpoch(20)
+	sectorExpiry := abi.ChainEpoch(100)
+
+	// caller is not the provider
+	{
+		t.Run("fail when caller is not the provider of the deal", func(t *testing.T) {
+			rt, actor := basicMarketSetup(t, marketActor, owner, provider, worker, client)
+			provider2 := tutil.NewIDAddr(t, 201)
+			mAddrs2 := &minerAddrs{owner, worker, provider2}
+			dealId := actor.generateAndPublishDeal(rt, client, mAddrs2, startEpoch, endEpoch)
+
+			params := mkActivateDealParams(sectorExpiry, dealId)
+
+			rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+			rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+			rt.ExpectAbort(exitcode.ErrIllegalState, func() {
+				rt.Call(actor.ActivateDeals, params)
+			})
+
+			rt.Verify()
+		})
+	}
+
+	// caller is not a StorageMinerActor
+	{
+		t.Run("fail when caller is not a StorageMinerActor", func(t *testing.T) {
+			rt, actor := basicMarketSetup(t, marketActor, owner, provider, worker, client)
+			rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+			rt.SetCaller(provider, builtin.AccountActorCodeID)
+			rt.ExpectAbort(exitcode.ErrForbidden, func() {
+				rt.Call(actor.ActivateDeals, &market.ActivateDealsParams{})
+			})
+
+			rt.Verify()
+		})
+	}
+
+	// deal has not been published before
+	{
+		t.Run("fail when deal has not been published before", func(t *testing.T) {
+			rt, actor := basicMarketSetup(t, marketActor, owner, provider, worker, client)
+			params := mkActivateDealParams(sectorExpiry, abi.DealID(42))
+
+			rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+			rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+			rt.ExpectAbort(exitcode.ErrIllegalState, func() {
```

Comment:
This exitcode seems unexpected.  I think it comes from the RequireNoErr after ValidateDealsForActivation.  That should probably be changed to `ErrIllegalArgument`

Reply:
Thinking more it looks a little more complicated.  Some failures from that function are IllegalState, some are IllegalArgument

Reply:
This is a known problem. I've filed an issue for it. https://github.com/filecoin-project/specs-actors/issues/597